src/computability/regex.lean

```diff
@@ -0,0 +1,216 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Fox Thomson.
+-/
+import data.fintype.basic
+import data.finset.basic
+import tactic.rcases
+import tactic.omega
+
+/-!
+# Regular Expression
+
+This file contains the formal definition for regular expressions and basic lemmas.
+-/
+
+universe u
+
+variables {α : Type u} [dec : decidable_eq α]
+
+/-- This is the definition of regular expressions
+  `RZero` matches nothing
+  `RNull` matches only the empty string
+  `RChar a` matches only the string 'a'
+  `RStar M` matches any finite concatenation of strings which match `M`
+  `RPlus M N` matches anything which match `M` or `N`
+  `RComp M N` matches `x ++ y` if `x` matches `M` and `y` matches `N` -/
+inductive regex (α : Type u) : Type (u+1)
+| RZero : regex
+| RNull : regex
+| RChar : α → regex
+| RStar : regex → regex
+| RPlus : regex → regex → regex
+| RComp : regex → regex → regex
+
+namespace regex
+
```

Comment:
Is it sensible to add something like the following?
```lean
instance : has_add (regex α) := ⟨RPlus⟩
instance : has_mul (regex α) := ⟨RComp⟩
```
Unfortunately, you'll find when you try to prove `(a + b) + c = a + (b + c)`, as the page on Kleene algebras suggests is sensible notation, that you can't.

If you instead work with `quot (\lam r s, \forall l, r.rmatch l = s.rmatch l)`, then you can prove that lemma.